llvm/lib/IR/RepoDefinition.cpp

```diff
@@ -119,7 +119,7 @@ void ModuleHashGenerator::calculateGOInfo(const GlobalType *G) {
     assert(isa<GlobalVariable>(GO) &&
            "Contributions must be global variables!");
     GOIMap[GO].Dependencies.emplace_back(G);
-    GInfo.Dependencies.emplace_back(GO);
+    GOIMap[G].Dependencies.emplace_back(GO);
```

Comment:
This change is a little perplexing! Superficially, G is invariant in this loop, so GOMap[G] should be the same. However… I can see that GOIMap is an LLVM DenseMap<> which doesn’t preserve iterators on insertion. That then suggests that the bug occurs when there was no entry in GOIMap for GO. Is this correct?

My concern here is that GOIMap[x] does a search of the hash table. We obviously want to avoid doing that search twice for every iteration of this loop. Instead of searching every time could we detect this condition and only perform the search when necessary?

Reply:
After writing this I saw your comment [here](https://github.com/SNSystems/llvm-project-prepo/issues/130#issuecomment-733814795) which answers some of my questions.

Reply:
> Instead of searching every time could we detect this condition and only perform the search when necessary?

Here’s my proposal to do exactly that. Note that I've made a couple of other tweaks to update comments that I found confusing, add const-ness, tweak names, etc. I've removed the assertion that contributions must be global variables: this function doesn't care or need to know (and it may change)!

~~~cpp
template <typename GlobalType>
void ModuleHashGenerator::calculateGOInfo(const GlobalType *const G) {
  const GOInfo GInitInfo = calculateDigestAndDependenciesAndContributions(G);
#ifndef NDEBUG
  LLVM_DEBUG(dbgs() << "\nGO Name:" << G->getName() << "\n");
  LLVM_DEBUG(GInitInfo.dump(true));
#endif
  GOInfo *GInfo = &GOIMap[G];

  GInfo->InitialDigest = GInitInfo.InitialDigest;

  // Append G's initial dependencies.
  std::copy(GInitInfo.Dependencies.begin(), GInitInfo.Dependencies.end(),
            std::back_inserter(GInfo->Dependencies));

  // The contributions of an object represent a bi-directional link to the
  // contained objects. Here we convert each contribution into entries in the
  // dependency lists for the two objects.
  for (auto &Contribution : GInitInfo.Contributions) {
    const std::pair<GOInfoMap::iterator, bool> EmplaceResult =
        GOIMap.try_emplace(Contribution);
    // If we just created an entry in GOIMap for GO, we invalidated GInfo so
    // must re-seat it by finding G once again.
    if (EmplaceResult.second) {
      GOInfoMap::iterator NewGPos = GOIMap.find(G);
      assert(NewGPos != GOIMap.end() &&
             "There must be an entry for G in GOIMap");
      GInfo = &NewGPos->second;
    }
    EmplaceResult.first->second.Dependencies.emplace_back(G);
    GInfo->Dependencies.emplace_back(Contribution);
  }
}
~~~

Reply:
Thanks Paul, I have corrected it in the commit https://github.com/SNSystems/llvm-project-prepo/pull/131/commits/a3e3e2ff581cdc8c0494f009309aa6c92096a8dd.